keyboards/bminiex/backlight.c

```diff
@@ -0,0 +1,213 @@
+/**
+ * Backlighting code for PS2AVRGB boards (ATMEGA32A)
+ * Kenneth A. (github.com/krusli | krusli.me)
+ */
+
+#include "backlight.h"
+#include "quantum.h"
+
+#include <avr/pgmspace.h>
+#include <avr/interrupt.h>
+
+#include "backlight_custom.h"
+#include "breathing_custom.h"
+
+// DEBUG
+#include <stdlib.h>
+#include <stdio.h>
+
+// Port D: digital pins of the AVR chipset
+#define NUMLOCK_PORT    (1 << 0)  // D0
+#define CAPSLOCK_PORT   (1 << 1)  // D1
+#define BACKLIGHT_PORT  (1 << 4)  // D4
+#define SCROLLLOCK_PORT (1 << 6)  // D6
+
+#define TIMER_CLK_DIV64			  0x03	///< Timer clocked at F_CPU/64
+#define TIMER1PRESCALE	TIMER_CLK_DIV64 ///< timer 1 prescaler default
+
+#define TIMER_PRESCALE_MASK		0x07	///< Timer Prescaler Bit-Mask
+
+#define PWM_MAX 0xFF
+#define TIMER_TOP 255 // 8 bit PWM
+
+extern backlight_config_t backlight_config;
+
+/**
+ * References
+ * Port Registers: https://www.arduino.cc/en/Reference/PortManipulation
+ * TCCR1A: https://electronics.stackexchange.com/questions/92350/what-is-the-difference-between-tccr1a-and-tccr1b
+ * Timers: http://www.avrbeginners.net/architecture/timers/timers.html
+ * 16-bit timer setup: http://sculland.com/ATmega168/Interrupts-And-Timers/16-Bit-Timer-Setup/
+ * PS2AVRGB firmware: https://github.com/showjean/ps2avrU/tree/master/firmware
+ */
+
+// @Override
+// turn LEDs on and off depending on USB caps/num/scroll lock states.
+__attribute__ ((weak))
+void led_set_user(uint8_t usb_led) {
+    if (usb_led & (1 << USB_LED_NUM_LOCK)) {
+      // turn on
+      DDRD  |= NUMLOCK_PORT;
+      PORTD |= NUMLOCK_PORT;
+    } else {
+      // turn off
+      DDRD  &= ~NUMLOCK_PORT;
+      PORTD &= ~NUMLOCK_PORT;
+    }
+
+    if (usb_led & (1 << USB_LED_CAPS_LOCK)) {
+      DDRD  |= CAPSLOCK_PORT;
+      PORTD |= CAPSLOCK_PORT;
+    } else {
+      DDRD  &= ~CAPSLOCK_PORT;
+      PORTD &= ~CAPSLOCK_PORT;
+    }
+
+    if (usb_led & (1 << USB_LED_SCROLL_LOCK)) {
+      DDRD  |= SCROLLLOCK_PORT;
+      PORTD |= SCROLLLOCK_PORT;
+    } else {
+      DDRD  &= ~SCROLLLOCK_PORT;
+      PORTD &= ~SCROLLLOCK_PORT;
+    }
+}
+
+#ifdef BACKLIGHT_ENABLE
+
+// sets up Timer 1 for 8-bit PWM
+void timer1PWMSetup(void) { // NOTE ONLY CALL THIS ONCE
```

Comment:
You wouldn't have to do that at all if #3615 were merged :) including breathing.
This code is in lots of different keyboards in QMK unfortunately...